File: leetcode/easy/844-backspaceStringCompare.js
// Given two strings S and T, return if they are equal when both are typed into empty text editors. # means a backspace character.

// Note that after backspacing an empty text, the text will continue empty.

// Example 1:

// Input: S = "ab#c", T = "ad#c"
// Output: true
// Explanation: Both S and T become "ac".
// Example 2:

// Input: S = "ab##", T = "c#d#"
// Output: true
// Explanation: Both S and T become "".
// Example 3:

// Input: S = "a##c", T = "#a#c"
// Output: true
// Explanation: Both S and T become "c".

// first pass solution
var backspaceCompare = function(s, t) {
    let stack1 = [];
    let stack2 = [];
    
    for (let char of s) {
        if (char === "#") stack1.pop();
        else stack1.push(char);
    }
    
    for (let char of t) {
        if (char === "#") stack2.pop();
        else stack2.push(char);
    }
    return stack1.join("") === stack2.join("");
};

File: leetcode/easy/844-backspaceStringCompare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        Stack<Character> stack1 = new Stack<>();
        Stack<Character> stack2 = new Stack<>();
        
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '#' && !stack1.isEmpty()) {
                stack1.pop();
            } else if (s.charAt(i) != '#') {
                stack1.push(s.charAt(i));
            }
        }
        
        for (int i = 0; i < t.length(); i++) {
            if (t.charAt(i) == '#' && !stack2.isEmpty()) {
                stack2.pop();
            } else if (t.charAt(i) != '#') {
                stack2.push(t.charAt(i));
            }
        }
      
        if (stack1.size() != stack2.size()) return false;
        
        while (stack1.size() > 0) {
            if (stack1.pop() != stack2.pop()) {
                return false;
            }
        }
        return true;
    }
}
